### DISCUSSION TITLE
How to import spaCy’s types avoiding the loading delay?

### DISCUSSION THREAD

**Discussion by danmysak** [explosion/spaCy#8986] [Help: Best practices]
Hi,

Let’s say I’m using spaCy conditionally. Let’s also say I want to decrease my script’s loading time by importing the library only when it is actually required. All is good so far: I can simply put `import spacy` into a function that gets called when the library is needed. But if the function returns a result of type, say, `spacy.language.Doc`, how do I annotate it without loading the module first? Both `import spacy` and `import spacy.language` take a couple of seconds to execute.

Thanks to all the maintainers for the great tool, by the way!

**Comment by adrianeboyd** [comment]
With the caveat that I'm not an expert on this, I think you can use an `if typing.TYPE_CHECKING` block to only import if needed for type checking, and then annotate the types as strings. We only do this to avoid circular imports, but I think it would be the same to avoid loading if not needed.

**Reply by danmysak** [reply]
Thanks!

I never used `TYPE_CHECKING` before. It seems to do the job, and in my case there was actually no need to change any types to strings—although the code looks somewhat quirky:

```
if TYPE_CHECKING:
    from spacy.language import Doc, Language
    from spacy.tokens import Token
else:
    Doc = Language = Token = NoReturn
```

**Reply by danmysak** [reply]
It’s a real pity though that the same approach cannot be used for `spacy.symbols`.
